Sources/web3swift/Contract/EthereumContract.swift

```diff
@@ -6,137 +6,140 @@
 import Foundation
 import BigInt
 
-public struct EthereumContract: ContractProtocol {
+public class EthereumContract: ContractProtocol {
+
     public var transactionOptions: TransactionOptions? = TransactionOptions.defaultOptions
     public var address: EthereumAddress? = nil
 
-    var _abi: [ABI.Element]
+    public let abi: [ABI.Element]
 
-    public var allEvents: [String] {
-        return events.keys.compactMap({ (s) -> String in
-            return s
-        })
-    }
+    private(set) public lazy var methods: [String: [ABI.Element.Function]] = {
+        var methods = [String: [ABI.Element.Function]]()
 
-    public var allMethods: [String] {
-        return methods.keys.compactMap({ (s) -> String in
-            return s
-        })
-    }
+        func appendFunction(_ key: String, _ value: ABI.Element.Function) {
+            var array = methods[key] ?? []
+            array.append(value)
+            methods[key] = array
+        }
 
-    public struct EventFilter {
-        public var parameterName: String
-        public var parameterValues: [AnyObject]
-    }
+        for case let .function(function) in abi where function.name != nil {
+            appendFunction(function.name!, function)
+            appendFunction(function.signature, function)
+            appendFunction(function.methodString.addHexPrefix().lowercased(), function)
 
-    public var methods: [String: ABI.Element] {
-        var toReturn = [String: ABI.Element]()
-        for m in self._abi {
-            switch m {
-            case .function(let function):
-                guard let name = function.name else {continue}
-                toReturn[name] = m
-            default:
-                continue
+            /// ABI cannot have two functions with exactly the same name and input arguments
+            if (methods[function.signature]?.count ?? 0) > 1 {
+                fatalError("Given ABI is invalid: contains two functions with possibly different return values but exactly the same name and input parameters!")
```

Comment:
Not my story, maybe @mloit have something to say?

Reply:
I'll consider implementing this check within `init` and maybe declaring this variable, and other similar ones, as `let` instead.
This way we can return `nil` from initializers or throw an error with a detailed description.

Reply:
Speaking about a let one, private variable required here with computed wrapper, because in reference types let will not guarantee constancy of value under it, it'll just guarantee the constancy of reference to it, so it could be changed occasionally by a user.

Reply:
I think it doesn't apply in our case and here's why:
- I've marked all variables as `public let` except `transactionOptions` and `address`. All collections are immutable now. 
- All types used in the following collections are `struct`s: `abi`, `methods`, `allMethods`, `events`, `allEvents`.  `constructor` is also struct. Thus, if anyone extracts it to a variable it will be a full copy. 
- Moreover - all types under `ABI.Element` have their internal variables declared as `let` which is good and must stay this way until web3 goes extinct.

So to recap - I do not see the actual issue here with marking variables as `public let`. If I didn't get the idea of your message let's discuss it.